### PR TITLE
Better error for missing tuple pattern in args

### PR DESCRIPTION
**Pull request by sinkuu**
#44150

Before:
```
error[E0593]: closure takes 2 arguments but 1 argument is required
 --> test.rs:5:40
  |
5 |     let it = v.into_iter().enumerate().map(|i, x| i);
  |                                        ^^^ -------- takes 2 arguments
  |                                        |
  |                                        expected closure that takes 1 argument
```

After:
```
error[E0593]: closure takes 2 arguments but a 2-tuple is required
 --> test.rs:5:40
  |
5 |     let it = v.into_iter().enumerate().map(|i, x| i);
  |                                        ^^^ ------ takes 2 arguments
  |                                        |
  |                                        expected closure that takes a 2-tuple
```

### CODE REVIEW

File: src/librustc/traits/error_reporting.rs [hunk 1]
```diff
@@ -711,41 +711,77 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
...
+                                Some(tuptys.len())
+                            } else {
+                                None
+                            }
+                        })
+                    } else {
+                        None
+                    };
+
+                    // FIXME(#44150): Expand this to "N args expected but a N-tuple found."
```
Comment (nikomatsakis) [inline src/librustc/traits/error_reporting.rs]: Why does this say "FIXME"?

**Comment by nikomatsakis** [comment]
This seems like a great thing to fix. I am wondering though if we should tweak the output. I personally found the new message a bit hard to parse, and I think it could be confusing. Instead of changing the actual text of the error message, I wonder if it'd be better to offer a suggestion. I am imagining something like this:

- In the event that N parameters were found
- But 1 parameter was expected, and it is a tuple of arity N
- Then offer a suggestion like "a N-tuple was expected, consider changing the closure signature like this: `|(a, b)|`

I imagine it looking something like this in the end:

```
error[E0593]: closure takes 2 arguments but 1 argument is required
 --> test.rs:5:40
  |
5 |     let it = v.into_iter().enumerate().map(|i, x| i);
  |                                        ^^^ ------ takes 2 arguments
  |                                        |
  |                                        expected closure that takes 1 argument
  = help: consider changing the closure signature like this:
5 |     let it = v.into_iter().enumerate().map(|(i, x|) i);
```

Thoughts @estebank?

To do this, you would basically invoke `span_suggestion` on the `DiagnosticBuilder` returned by `arg_count_mismatch` (or at the end of that function).

We could also keep things just as they are, but *also* include the suggestion. I'm not sure which is better, but I definitely the suggestion is a good idea, since I think it will have a more obvious meaning to people than talking about 2-tuples.

**Comment by nikomatsakis** [comment]
Another option is to do something more dramatic:

```rust
error[E0593]: closure takes multiple arguments, but a tuple argument is required
 --> test.rs:5:40
  |
5 |     let it = v.into_iter().enumerate().map(|i, x| i);
  |                                        ^^^ ------ consider changing to `|(i, x)|`
  |                                        |
  |                                        expected closure that takes 1 argument, a 2-tuple
```

File: src/librustc/traits/error_reporting.rs (hunk 1 above)
Comment (sinkuu) [reply src/librustc/traits/error_reporting.rs]: This PR improves message for "expected |(x, y)|, found |x, y|" case but not "expected |x, y|, found |(x, y)|" case. In the latter case, `found_trait_ref` should be `FnMut<((_, _),)>`, but it become `FnMut<(_,)>` and I don't know how to fix.

**Comment by sinkuu** [comment]
Added suggestions.

File: src/librustc/traits/error_reporting.rs [hunk 2]
```diff
@@ -767,32 +831,87 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
...
-            if expected == 1 { "" } else { "s" },
+            "{} takes {}, but {} {} required",
+            kind,
+            if expected_tuple.is_some() {
+                Cow::from("multiple arguments")
+            } else {
+                Cow::from(args_str(found))
+            },
+            if expected_tuple.is_some() {
+                Cow::from("a tuple argument")
```
Comment (nikomatsakis) [inline src/librustc/traits/error_reporting.rs]: I think I figured out what is bothering me about this wording. The construction is not quite parallel. It says "multiple arguments" are expected, but then says "but a tuple argument", which is not obviously the opposite. Also, our usual construction is "expected X, but found Y", whereas the message here says "found X, but Y is required" (pre-existing, I know, but I still find it a bit surprising).

Perhaps this?

**expected a closure that takes a single tuple as argument but found a closure that takes N distinct arguments**

Maybe that's too wordy. We could tweak it, though it goes a *bit* away from the "expected X but found Y" form:

**expected closure to take a single tuple as argument but it takes N distinct arguments**

or, better:

**closure is expected to take a single tuple as argument, but it takes N distinct arguments**

(I would probably change the existing message that doesn't mention tuples to match, as well.)

File: src/test/ui/mismatched_types/closure-arg-count.stderr [hunk 3]
```diff
@@ -23,23 +23,47 @@ error[E0308]: mismatched types
...
+20 |     let _it = vec![1, 2, 3].into_iter().enumerate().map(|i, x| i);
+   |                                                     ^^^ ------ help: consider changing to: `|(i, x)|`
+   |                                                     |
+   |                                                     expected closure that takes 1 argument, a 2-tuple
+
+error[E0593]: closure takes multiple arguments, but a tuple argument is required
+  --> $DIR/closure-arg-count.rs:21:53
+   |
+21 |     let _it = vec![1, 2, 3].into_iter().enumerate().map(|i: usize, x| i);
+   |                                                     ^^^ ------------- help: consider changing to: `|(i, x): (usize, _)|`
```
Comment (nikomatsakis) [inline src/test/ui/mismatched_types/closure-arg-count.stderr]: this looks great

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This is awesome! What do you think about my suggest changes to the error message wording?

cc @estebank -- this is definitely the kind of thing we should have a style guide for!

File: src/librustc/traits/error_reporting.rs (hunk 2 above)
Comment (sinkuu) [reply src/librustc/traits/error_reporting.rs]: It looks better, thanks! Reworded the message to the last one.

File: src/test/ui/mismatched_types/closure-arg-count.stderr [hunk 4]
```diff
@@ -23,23 +23,47 @@ error[E0308]: mismatched types
...
    = note: required by `f`
 
-error: aborting due to 5 previous errors
+error[E0593]: closure is expected to take a single tuple as argument, but it takes 2 distinct arguments
+  --> $DIR/closure-arg-count.rs:20:53
+   |
+20 |     let _it = vec![1, 2, 3].into_iter().enumerate().map(|i, x| i);
+   |                                                     ^^^ ------ help: consider changing to: `|(i, x)|`
+   |                                                     |
+   |                                                     expected closure that takes 1 argument, a 2-tuple
```
Comment (nikomatsakis) [inline src/test/ui/mismatched_types/closure-arg-count.stderr]: This is looking really good! I hope you don't mind me nitpicking some more though. =)

Reading this "expected closure that takes 1 argument, a 2-tuple" -- which I think I may even have initially suggested -- I see now that it is actually ambiguous. In particular, the intention was something like "that takes 1 argument, which is a 2-tuple", but it could also be parsed as a list, like "that takes 1 argument and a 2-tuple". In fact, this second interpretation is what I got this time, and I had to re-read it a few times to understand what was meant.

Therefore, I suggest:

"expected closure that takes a single tuple as argument"

which echoes the wording from above

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc/traits/error_reporting.rs (hunk 1 above)
Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: @nikomatsakis, I believe @sinkuu is referring to the following case:

```
 18 |     f(|(x, y)| x + y);
    |     ^ -------- takes a single tuple as an argument
    |     | |
    |     | help: consider changing to: `|x, y|`
    |     |
    |     expected closure that takes 2 arguments
```

We should create an issue after this PR lands to track that case.

File: src/test/ui/mismatched_types/closure-arg-count.stderr [hunk 5]
```diff
@@ -23,23 +23,47 @@ error[E0308]: mismatched types
...
    |     expected closure that takes 1 argument
    |
    = note: required by `f`
 
-error: aborting due to 5 previous errors
+error[E0593]: closure is expected to take a single tuple as argument, but it takes 2 distinct arguments
+  --> $DIR/closure-arg-count.rs:20:53
+   |
+20 |     let _it = vec![1, 2, 3].into_iter().enumerate().map(|i, x| i);
+   |                                                     ^^^ ------ help: consider changing to: `|(i, x)|`
```
Comment (estebank) [inline src/test/ui/mismatched_types/closure-arg-count.stderr]: Could you modify the suggestion text to be freestanding without the snippet? Something along the lines of `consider changing the closure to accept a tuple`. `rustc` is not the only way these messages can be presented, so they should make even when the suggested code is not presented next to it.

**Review by estebank — CHANGES_REQUESTED** [review]

**Review by nikomatsakis — APPROVED** [review]

**Review by estebank — APPROVED** [review]

**Comment by ExpHP** [comment]
Seeing this error message actually brightened up my day.  So many times have I made this mistake... 
 Thanks for adding this.
